dmutils/direct_plus_client.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import requests
+
+
+class DirectPlusClient(object):
+    """Client to interface with Dun and Bradstreet's Direct Plus API."""
+
+    access_token = None
+    protocol = 'https'
+    domain = 'plus.dnb.com'
+    required_headers = (
+        ('Content-Type', 'application/json'),
+        ('Accept', 'application/json'),
+    )
+
+    def __init__(self, username, password, logger=None):
+        self.username = username
+        self.password = password
+        self.logger = logger if logger else logging.getLogger(__name__)
+
+    def _reset_access_token(self):
+        """
+        Set the access token and header authorisation parameter for requests to use.
+        """
+        basic_auth_string = requests.auth._basic_auth_str(self.username, self.password)
+        response = self._direct_plus_request(
+            'token',
+            method='post',
+            version='v2',
+            payload={'grant_type': 'client_credentials'},
+            extra_headers=(('Authorization', basic_auth_string),),
+            allow_access_token_reset=False
+        )
+        self.access_token = response.json().get('access_token')
+        self.required_headers += (('Authorization', f'Bearer {self.access_token}'),)
+
+    def _direct_plus_request(
+        self,
+        endpoint,
+        method='get',
+        version='v1',
+        payload=None,
+        extra_headers=(),
+        allow_access_token_reset=True
+    ):
+        """
+        Make a request to the Direct Plus API
+        """
+        if self.access_token is None and allow_access_token_reset is True:
+            self._reset_access_token()
+
+        url = f'{self.protocol}://{self.domain}/{version}/{endpoint}'
+        headers = {**dict(self.required_headers), **dict(extra_headers)}
+
+        if method != 'get':
+            response = getattr(requests, method)(url, headers=headers, json=payload)
+        else:
+            response = requests.get(url, headers=headers, params=payload)
+
+        if response.status_code == 401 and allow_access_token_reset is True:
+            # If access token invalid (401) refresh access token manually
+            # and retry request with allow_access_token_reset = False
+            self._reset_access_token()
+            response = self._direct_plus_request(
+                endpoint,
+                method=method,
+                version=version,
+                payload=payload,
+                extra_headers=extra_headers,
+                allow_access_token_reset=False
+            )
+        return response
+
+    def get_organization_by_duns_number(self, duns_number):
+        """
+        Request a supplier by duns number from the Direct Plus API
+        """
+        response = self._direct_plus_request(
+            f'data/duns/{duns_number}', payload={'productId': 'cmpelk', 'versionId': 'v2'}
```

Comment:
Where is `cmpelk` coming from? Worth hiving this out into a constant at the top of the file in case it changes?

Reply:
https://directplus.documentation.dnb.com/html/resources/JSONSample_CMP.html#cmpelk

`CMPELK` is an internal D&B code that sort of references the format of the response.

They call it a product code. So we're getting an 'organization' object from Dun and Bradstreets 'CMPELK' product using their 'Direct Plus API' interface.

But it's specific to getting a supplier org from this endpoint rather than referencing the API as a whole.

Reply:
Makes sense, thanks. And now we have it documented in this PR 😸